### ISSUE TITLE
Authentication / Authorization

### ISSUE THREAD

**Issue by atalebagha** [nodkz/graphql-compose#29]
How do I add authentication/authorization/permissions to queries and mutations?

example, User can view all users but can only edit his user account.
User who is admin of a business can add/remove users from the business and can make/receive payments
User who is an employee of a business can view payments made to/from the business

**Comment by stoffern** [comment]
It is easier if you provide your code.
First off you need to get user ID and pass it into GraphQL if authenticated.

server.js
```js
app.use('/graphql', 
  graphqlHTTP(req => ({
    schema: Schema,
    graphiql: true,
    context: {
      userId: req.user._id,
    },
  }))
)
```

Users.js
```js
var UserSchema = new mongoose.Schema({
  email : { type: String },
  firstName : { type: String },
  lastName : { type: String },
})
export const Users = mongoose.model('Users', UserSchema)
export const UsersTC = composeMongoose(Users)
```

schema.js
```js
export const RootQueryTC = GQC.rootQuery()
export const RootMutationTC = GQC.rootMutation()

RootQueryTC.addFields({
  users: UsersTC.get('$findMany')
})

RootMutationTC.addFields({
  userUpdate: UsersTC.get('$updateById')
    .wrapResolve( next => req => {
      if(req.context.userId != req.args.input.record._id)
        throw new Error ('Not allowed')
      return next(req)
    }),

  addUserToBusiness: BusinessTC.get('$updateById')
    .wrapResolve( next => req => {
        UsersTC.find({_id: req.context.userId}).exec( user => {
          //if not owner throw Error
          //if owner and you want to parse data, change the req data
        })
      return next(req)
    }),
})

export default = GQC.buildSchema()
```

**Comment by nodkz** [comment]
Other beauty way how to restrict access is 
```js
import { GQC } from 'graphql-compose';
import { MailTC } from './mail';
import { CompanyTC } from './company';
import { NoteTC } from './note';
import { UserTC } from './user';

export const RootMutationTC = GQC.rootMutation();

RootMutationTC.addFields({
  companyCreate: CompanyTC.get('$createOne'), // may anybody
  noteCreate: NoteTC.get('$createOne'),  // may anybody

  ...adminAccess({
    userCreate: UserTC.get('$createOne'),
    userUpdate: UserTC.get('$updateById'),
    userRemove: UserTC.get('$removeById'),
    sendMail: MailTC.get('$sendMail'),
  }),
});

function adminAccess(resolvers) {
  Object.keys(resolvers).forEach((k) => {
    resolvers[k] = resolvers[k].wrapResolve(next => (rp) => {
      if (!rp.context.admin) {
        throw new Error('You should be admin, to have access to this action.');
      }
      return next(rp);
    });
  });
  return resolvers;
}
```

**Comment by atalebagha** [comment]
Perfect! The above looks like it will work great! Thank you

Here's another question with regards to field access:

```javascript
var UserSchema = new mongoose.Schema({
  email : { type: String },
  firstName : { type: String },
  lastName : { type: String },
  role: { type: String },
  businessRole: { type: String },
  businessId: { type: ObjecId, ref: 'Business' },
  isAdminLocked: { type: Boolean },  // only a super user can change this value, (outside of business)
  isLocked: { type: Boolean } // only a business admin from the same business can change/view this
})
export const Users = mongoose.model('Users', UserSchema)
export const UsersTC = composeMongoose(Users)
```

How about field permissions for viewing and changing.  The above fields, I want it hidden and non-writeable from non-admin users.  The first one, is the application admin, and the second is just a business admin who is from the same business as the user.

Thank you!

**Comment by stoffern** [comment]
Permission for changing is covered in the two posts in top.
Permission for viewing can be solved using projections..

```js
ViewerTC.addRelation(
  'currentUser',
  () => ({
    resolver: UsersTC.getResolver('findById'),
    args: {
       _id: (source, args, context, info) => context.userId,
    },
  })
)
```

```js

ViewerTC.addRelation(
  'userSearch',
  () => ({
    resolver: UsersTC.getResolver('findMany'),
    args: {
      filter: (source) => ({ email: source.email }),
      sort: null,
    },
    projection: { businessRole: false, businessId: false },
  })
)
```

**Comment by atalebagha** [comment]
Thanks @stoffern 

To add to the above use case, what if I want the viewing of that field is dependent on the user context?

Example:  user with admin role can see businessRole.  How do I make projections dynamic based upon the context?

**Comment by nodkz** [comment]
@stoffern @atalebagha  I will not connect projections with context. Let's keep it simple.
To solve your case better will be to use standard graphql `resolve` method for fields.

Right now published new package `graphql-compose@1.4.4` with `TC.extendField(name, fieldConfig)` method.
You may add your logic to schema in such way:
```js
UsersTC.extendField('businessRole', {
  description: 'May see only admins',
  resolve: (source, args, context) => (context.isAdmin ? source.businessRole : null),
});
```

**Comment by atalebagha** [comment]
thanks @nodkz
